libs/telemetry/src/test/java/org/opensearch/telemetry/tracing/DefaultTracerTests.java

```diff
@@ -104,51 +119,239 @@ public void testCreateSpanWithParent() {
+     * 4. Scope.close
+     * 5. Parent Scope.close
+     * 6. verify the current_span is still the same on async thread as the 2
+     * 7. verify the main thread has current span as null.
+     */
+    public void testSpanAcrossThreadsMultipleSpans() {
+        TracingTelemetry tracingTelemetry = new MockTracingTelemetry();
+        ThreadContext threadContext = new ThreadContext(Settings.EMPTY);
+        AtomicReference<Span> currentSpanRefThread1 = new AtomicReference<>();
+        AtomicReference<Span> currentSpanRefThread2 = new AtomicReference<>();
+        AtomicReference<Span> currentSpanRefAfterEndThread2 = new AtomicReference<>();
+
+        AtomicReference<Span> parentSpanRef = new AtomicReference<>();
+        AtomicReference<Span> spanRef = new AtomicReference<>();
+        AtomicReference<Span> spanT2Ref = new AtomicReference<>();
+
+        ThreadContextBasedTracerContextStorage spanTracerStorage = new ThreadContextBasedTracerContextStorage(
+            threadContext,
+            tracingTelemetry
+        );
+        DefaultTracer defaultTracer = new DefaultTracer(tracingTelemetry, spanTracerStorage);
+
+        executorService.execute(() -> {
+            // create a parent span
+            Span parentSpan = defaultTracer.startSpan(new SpanCreationContext("p_span_name", Attributes.EMPTY));
+            SpanScope parentSpanScope = defaultTracer.createSpanScope(parentSpan);
+            parentSpanRef.set(parentSpan);
+            // create a span
+            Span span = defaultTracer.startSpan(new SpanCreationContext("span_name_t_1", Attributes.EMPTY));
+            SpanScope spanScope = defaultTracer.createSpanScope(span);
+            spanRef.set(span);
+
+            executorService.execute(() -> {
+                Span spanT2 = defaultTracer.startSpan(new SpanCreationContext("span_name_t_2", Attributes.EMPTY));
+                SpanScope spanScopeT2 = defaultTracer.createSpanScope(spanT2);
+                spanT2Ref.set(spanT2);
+
+                Span spanT21 = defaultTracer.startSpan(new SpanCreationContext("span_name_t_2", Attributes.EMPTY));
+                SpanScope spanScopeT21 = defaultTracer.createSpanScope(spanT2);
+
+                currentSpanRefThread2.set(defaultTracer.getCurrentSpan().getSpan());
+
+                spanT21.endSpan();
+                spanScopeT21.close();
+
+                spanT2.endSpan();
+                spanScopeT2.close();
+                currentSpanRefAfterEndThread2.set(defaultTracer.getCurrentSpan() != null ? defaultTracer.getCurrentSpan().getSpan() : null);
+            });
+            spanScope.close();
+            parentSpanScope.close();
+            currentSpanRefThread1.set(defaultTracer.getCurrentSpan().getSpan());
+        });
+        assertEquals(spanT2Ref.get(), currentSpanRefThread2.get());
+        assertEquals(spanRef.get(), currentSpanRefAfterEndThread2.get());
+        assertEquals(null, currentSpanRefThread1.get());
     }
 
```

Comment:
Could you please add a test case for https://github.com/opensearch-project/OpenSearch/pull/9617/files#r1311957624 (when method being called returns `CompleteableFuture` fe) ? Or/and uses `ActionListener<Response>` ? Thank you.

Reply:
There is already one test written on these line - https://github.com/opensearch-project/OpenSearch/pull/9617/files#diff-6e9bf39593fc8bce77b5d3508e3174f69a2803fec0465f5dd46d736dfdb320ecR254 Please check if this is the same that you are looking for?

Reply:
> There is already one test written on these line - https://github.com/opensearch-project/OpenSearch/pull/9617/files#diff-6e9bf39593fc8bce77b5d3508e3174f69a2803fec0465f5dd46d736dfdb320ecR254 Please check if this is the same that you are looking for?

Sorry, the link doesn't work for me, do you mean `testSpanAcrossThreads` method in `DefaultTracerTests`? If yes - it does not exercise the flow when span is created in T1 but ended in T2.

Reply:
Yes, you are right. Adding one.